README.md
 $Id: README.md 1398 2021-07-08 10:47:54Z ltaulell $
 SPDX-License-Identifier: BSD-2-Clause

# PSMN ZFS toolbox

PSMN's python3 zfs toolbox is a rewrite of a unmaintenable set of 
bash scripts (gZFS).

## TL;DR

The idea is to have automated rolling snapshots (and replicas) in a time 
frame (default is one by day, rolling on seven days), using a set of 
python3 scripts.

It use ``execo`` python3 module (python3 -m pip install execo) from INRIA.

This is Work In Progress:
 * v0.1 tested (svn r837 - r859)
 * v0.2 tested (svn r916 - r921)
 * v0.3 tested (svn r935 - r936)
 * v0.4 tested (svn r946 - r950)
 * v0.5 tested (svn r1057 - r1397)
 * v0.6 tested (full git)
 * current dev

Successfully tested from Debian 9 to Debian 12.

In active production at PSMN and UMPA lab.

## Default directories and files (example)

```
/root
└── zfstoolbox
    ├── conf
    │   ├── replica-data-example.yml    (see zfs_conf_example.yml)
    │   ├── snapshot-data-example.yml   (see zfs_conf_example.yml)
    │   └── zfs_defaults.yml
    ├── manage_conf.py
    ├── zfs_actions.py                  (need exec bit)
    └── zfs_common.py

```

Of course, you may adapt it at your own needs, see below.

## default configuration file (zfs_defaults.yml, exemples)

```
%YAML 1.1
---
# crontab/ssh user
user: 'root'
# will be use for crontab creation date (empty)
date: ''
# same as __version__ in *.py
version: '0.6'
path:
  # main tools location
  tools: '/root/zfstoolbox/'
  # main script
  cmd: 'zfs_actions.py'
  # path to config dir (trailing / is mandatory)
  # home of zfs-defaults.yml
  conf: '/root/zfstoolbox/conf/'
  # for crontab, we need bash as shell, 'source' (for venv) beeing a builtin
  defaultshell: 'SHELL=/usr/bin/bash'
  # path to crontab dir (trailing / is mandatory)
  cron: '/etc/cron.d/'
  # path to log dir (trailing / is mandatory)
  log: '/var/log/'
  # path for zfs binary (autodetected?)
  zfs: ''
```

Be sure to modify the **CONFIGFILE** global variable in ``manage_conf.py`` and ``zfs_actions.py``, so that both scripts can read this configuration file.

## Requirements

```
Python >= 3.5 with standard installation (as in most distribution's packages)
execo >= 2.6.2
PyYAML >= 5.1.2
```

Obviously zfs, and ssh, if you do external replicas.

## VirtualEnv

As of Debian 12 'Bookworm', you need a virtual python environment:

### Maximum Debian

```
apt-get install python3-venv python3-yaml
python3 -m venv zfstoolbox --system-site-packages
cd zfstoolbox/ && source zfstoolbox/bin/activate
python3 -m pip install execo
```

### Maximum autonomous

```
apt-get install python3-venv
python3 -m venv zfstoolbox
cd zfstoolbox/ && source zfstoolbox/bin/activate
python3 -m pip install execo PyYAML
```

### Entering your virtual env

Copy files shown in example above, in the newly created hierarchy, with correct chmod.

Then enter your virtualenv, ``cd zfstoolbox/ && source bin/activate``, and start to work.

You can exit your virtualenv with: ``deactivate``.


## Create configuration files

For each automated snapshot/replica on a zfs dataset, a YAML configuration file
is needed: `manage_conf.py` helps create it.

### manage_conf.py

 * interactive
 * create the yaml config file
 * create the crontab file
 * restart cron for new crontab to be handled

```
usage:
  manage_conf.py {command} {type} <volume> [<crontab options> <args>]

Enable or disable a configuration for automated snapshots or replicas.

Where {command} is:
  enable                enable a new configuration and activate it
  disable               or disable (and erase) an existing configuration

Where {type} is:
  --snapshot            configuration for snapshot(s) on <volume>
  --replica             or configuration for replica(s) of <volume>

<volume>                mandatory zfs volume name

<crontab options> are:
  -H, --hour            Hour parameter (mandatory for enable)
  -m, --minute          minute parameter (mandatory for enable)
  -DM, --dayofmonth     Day of Month parameter
  -M, --month           Month parameter
  -DW, --dayofweek      Day of Week parameter

<args> can be:
  -d, --debug           toggle debug ON
  -l, --log             toggle log ON
  -r, --retention RETENTION
                        retention in days [default: 7]
  --mail [MAIL]         let cron send an email for each snapshot/replica
                        (default: no, [default: root])

For a replica, there are more optional <args>:
  -t, --target TARGET   zfs target pool name [default: None, same as source]
  -s, --server SERVER   server hosting replication target [default: localhost]

  --simple              replica type: simple, only target pool and last dataset [default]
  --complete            or complete, target pool plus complete source dataset'

  --latest              replica archive: only latest snapshot [default]
  --oldest              or complete set of snapshots

positional arguments:
  command     enable or disable

optional arguments:
  -h, --help  show this help message and exit

```
#### enabling/disabling snapshots

**example**: `python3 manage_conf.py enable --snapshot data/volume -H 1 -m 0`

Will create a configuration yaml file, to do a snapshot at 01:00 everyday, 
for zfs dataset 'data/volume', with a retention of 7 days, without log, 
debug and recapitulative mail. Will also create the corresponding crontab, 
and can activate it by restarting cron service.

**example**: `python3 manage_conf.py enable --snapshot --log -r 15 -H 22 -m 15 -DW 0 data/volume --mail`

Will create a configuration yaml file, to do a snapshot at 22:15 every Sunday, 
for zfs dataset 'data/volume', with a retention of 15 days, with log, without 
debug, but with recapitulative mail (to system default, usualy root). Will also 
create the corresponding crontab, and can activate it by restarting cron service.

If you need to run snapshot more than one by day, you should create a simple
crontab file, then modify it to your needs (man crontab), then restart 
cron service.

Be aware that zfs_actions.py destroy snapshot on a daily basis, hence all 
snapshots created the same day, will be destroyed **snapshot day + retention**
days later.

It might be necessary for the firsts runs to activate the mail option. 
After some successfull tests, you can add `>/dev/null` manually in the 
crontab.

**example**: `python3 manage_conf.py disable --snapshot data/volume`

Will erase crontab and configuration file for zfs volume 'data/volume', then
restart cron service, and propose to erase existing YAML configuration file. 
Remaining snapshot(s) will have to be destroy manually.

#### enabling/disabling replicas

**example**: `python3 manage_conf.py enable --replica data/volume -t backup -H 1 -m 0`

Will create a configuration yaml file, to do a replica at 01:00 everyday, 
of zfs dataset 'data/volume', to zfs dataset 'backup/volume' with a 
retention of 7 days, without log, debug and recapitulative mail, on
'localhost', simple and latest mode by default. 
Will also create the corresponding crontab, and can activate it by 
restarting cron service.

**example**: `python3 manage_conf.py enable --replica --log -s server2 -r 15 data/volume -H 23 -m 15 -DW 0`

Will create a configuration yaml file, to do a replica at 23:15 every Sunday, 
for zfs dataset 'data/volume', with a retention of 15 days, with log, without 
debug and recapitulative mail, on host 'server2', simple and latest mode by 
default. Will also create the corresponding crontab, and can activate it 
by restarting cron service.

If you need to run replica more than one by day, you should create a simple
crontab file, then modify it to your needs (man crontab), then restart 
cron service.

Be aware that zfs_actions.py destroy replica(s) on a daily basis, hence all 
replicas created the same day, will be erased **replica day + retention**
days later.

Please refer to ssh manual to create a ssh key pair (without passphrase).

**example**: `python3 manage_conf.py disable --replica data/volume`

Will erase crontab and configuration file of replica for zfs dataset 'data/volume',
then restart cron service. Remaining replica(s), snapshot(s) and dataset, will
have to be destroyed by hand on target hosting.


## Actual snapshots or replicas

### zfs_actions.py

 * executed by cron (or interactive)
 * read the yaml config file
 * create a snapshot (or send replica) as configured
 * if necessary, erase some snapshot(s) (or replica(s)), as configured

```
usage: zfs_actions.py [-h] -c CONF [-d]

Create, or replicate, and destroy automated snapshots

optional arguments:
  -h, --help            show this help message and exit
  -c CONF, --conf CONF  Mandatory YAML config file
  -d, --debug           toggle debug (default: no)
```

**example**: `zfs_actions.py -c /root/conf/snapshot-data-volume.yml`

Will create/destroy snapshots for zfs dataset 'data/volume', according to
configuration.

Be aware that zfs_actions.py erase snapshot on a daily basis, hence all 
snapshots created the same day, will be erased **snapshots day + retention**
days later.

**example**: `zfs_actions.py -c /root/conf/replica-data-volume.yml`

Will create/destroy replicas for zfs dataset 'data/volume', according to
configuration.

If target dataset doesn't exist, it will be created, according to configuration 
(simple/complete).

Be aware that zfs_actions.py erase replicas on a daily basis, hence all 
replicas created the same day, will be erased **replicas day + retention**
days later.

## Various / Bulk

  * logs

Logging use execo formatter, also in log files, which add colors with escape 
sequences. It may look weird with some pagers (less, without `export LESS=-ir`).

Logged interactive operations have a different color (orange) than 
automated operations (white).

  * automated log rotation

You may want to rotate your log, using logrotate. Create a file named 
'zfs_actions' in '/etc/logrotate.d/'.

```
# $CREATION_DATE
# /etc/logrotate.d/zfs_actions

/var/log/zfs_actions.log {
        su root adm
        monthly
        rotate 12
        missingok
        compress
        copytruncate
}
```

This example keep one year of logs, rotated monthly.

  * debug

Debug might be extremely verbose. We advise to not debug AND log 
automaticaly for a long period of time.

  * ansible

An ansible task, with templates, is available. See `ansible` directory.


  * bump version (prior to v0.4) to 0.4

``` bash
grep version conf/*.yml
sed -i '/version/ s/0.2/0.4/' conf/*.yml
```

  * bump version (prior to 0.3) to 0.5

``` bash
update_to_v0.5.py -f path/to/file.yml
```



CHANGELOG

## dev (ongoing development)

See TODO, readme-dev.fr.md

See existing branches

## v0.6

    Debian 12 is switching to full venv for unpackaged python3 tools:
        new YAML default configuration
            tools path and cmd are separated
            new defaultshell, for crontab
        manage_conf.py modified accordingly
        'version/__version__' changes in every file, including volume's configurations

## v0.5b (full git, no more subversion, no tag)

    add ansible task

## v0.5 (svn r1057, r1075, r1115)

    Code refactoring:
        one script to manage all confs (manage_conf.py)
        one script do performe action (zfs_actions.py)
        one module (zfs_common.py)
    v0.5 YAML default configuration and volume's configurations are incompatible
    with previous versions.
    A helper script can perform translation to new v0.5 options:
        (update_to_v0.5.py)
    Some sed oneliners are also provided for heavely diverging configs:
        (search_and_replace.sed)
    Updated README.md to v0.5
    #0002310@sourcesup: local replica => stderr, distant replica => stdout...
        merge stderr and stdout, refactor execute_zfs() (zfs_common.py)

## v0.4 (svn r941, r950)

    #0001938@sourcesup: zfs-replica does use nolog=True, if target dataset may 
        not exist hence, no WARNING in logs,
    add kwarg first=True/False on get_dist_snap (refactor),


## v0.3 (svn r934, r935, r939)

    *.py, *.yml bumped version to 0.3
    light refactoring in zfs-replica.py,
    enable-*.py add 'version' in conf files,
    version check added everywhere,
    documentation updated,
    #0001831@sourcesup: if target dataset does not exist, do a full replication 
        without WARNING.

WARNING! v0.3 scripts and configurations are NOT compatible with v0.2.


## v0.2 (svn r916, r918, r921)

    enable-replica.py, disable-replica.py and zfs-replica.py are usable,
  
    zfs-defaults.yml contain default configuration for both snapshot and 
    replica scripts. New config keys:
        path:
            snapshot
            replica
            conf
            log
            cron
            zfs: ''

        enable-snapshot.py, disable-snapshot.py and zfs-snapshot.py modified
        accordingly,

    zfs-conf-example.yml contain both examples for snapshot and replica 
    configurations,
    
    enable-*.py and disable-*.py have now a version check.

WARNING! v0.2 scripts and configurations are NOT compatible with v0.1.


## v0.1 (svn r837, r859)

    enable-snapshot.py, disable-snapshot.py and zfs-snapshot.py are usable,
    zfs-defaults.yml contain default configuration for python scripts,
    zfs-conf-example.yml contain one example of snapshot configuration,
    #0001749@sourcesup: dataset may contain a '-'.


zfstoolbox/manage_conf.py
#!/usr/bin/env python3
# coding: utf-8

# $Id: manage_conf.py 1398 2020-09-15 14:27:54Z gruiick $
# SPDX-License-Identifier: BSD-2-Clause

"""

    * manage (create/erase) config file for zfs volume provided by
    '<volume>', ask if it is correct before writing,
    * manage (create/erase) related crontab file, ask if it is correct
    before writing,
    * ask and restart cron.

TODO:

FIXME:

"""

import os
import sys
import argparse
from pprint import pprint
from datetime import datetime
import logging

import execo

import zfs_common

__version__ = "0.5"
__author__ = "See AUTHORS"
__copyright__ = "Copyright 2018, PSMN, ENS de Lyon"
__credits__ = "See CREDITS"
__license__ = "BSD-2"
__maintainer__ = "Loïs Taulelle"
__email__ = "None"
__status__ = "Production"

# global configuration file
CONFIGFILE = '/root/conf/zfs_defaults.yml'
# CONFIGFILE = 'zfs_defaults.yml'  # tests only


class ParseArgs(object):
    """ overcharge argparse and fill a dict of arguments, with defaults sets """

    def __init__(self):
        """ select according to first positional (act as mutually_exclusive_group) """
        self.arguments = {}
        parser = argparse.ArgumentParser(description='', usage='''
  %(prog)s {command} {type} <volume> [<crontab options> <args>]

Enable or disable a configuration for automated snapshots or replicas.

Where {command} is:
  enable                enable a new configuration and activate it
  disable               or disable (and erase) an existing configuration

Where {type} is:
  --snapshot            manage configuration file for snapshot(s) on <volume>
  --replica             or manage configuration file for replica(s) of <volume>

<volume>                mandatory zfs volume name

<crontab options> are:
  -H, --hour            Hour parameter (mandatory for enable)
  -m, --minute          minute parameter (mandatory for enable)
  -DM, --dayofmonth     Day of Month parameter
  -M, --month           Month parameter
  -DW, --dayofweek      Day of Week parameter

<args> can be:
  -d, --debug           toggle debug ON
  -l, --log             toggle log ON
  -r, --retention RETENTION
                        retention in days [default: 7]
  --mail [MAIL]         let cron send an email for each snapshot/replica
                        (default: no, [default: root])

For a replica, there are more optional <args>:
  -t, --target TARGET   zfs target pool name [default: None, same as source]
  -s, --server SERVER   server hosting replication target [default: localhost]

  --simple              replica type: simple, only target pool and last dataset [default]
  --complete            or complete, target pool plus complete source dataset'

  --latest              replica archive: only latest snapshot [default]
  --oldest              or complete set of snapshots

''')
        parser.add_argument('command', help='enable or disable')
        args = parser.parse_args(sys.argv[1:2])
        if not hasattr(self, args.command):
            print('Unrecognized command')
            parser.print_help()
            exit(1)
        # store everything in a global dict
        self.arguments.update(vars(args))
        # use dispatch pattern to invoke method with same name
        getattr(self, args.command)()

    def enable(self):
        """ Parse the enable part of arguments """
        # some default <args>
        enable_parser = argparse.ArgumentParser(description='enable a configuration',
                                                usage='%(prog)s enable {type} '
                                                '[options] <volume>', add_help=True)
        enable_parser.add_argument('-r', '--retention', nargs=1, type=int,
                                   help='retention, in day(s) [default: 7]')
        enable_parser.add_argument('--mail', action='store', nargs='?',
                                   type=str, const='root', help='let cron send '
                                   'an email each snapshot (default: no, '
                                   '[default: root])')
        # <crontab> options
        cron = enable_parser.add_argument_group('crontab options')
        cron.add_argument('-H', '--hour', nargs=1, type=int,
                          help='cron hour parameter', required=True)
        cron.add_argument('-m', '--minute', nargs=1, type=int,
                          help='cron minute parameter', required=True)
        cron.add_argument('-DM', '--dayofmonth', nargs=1, type=int,
                          help='cron day of month parameter')
        cron.add_argument('-M', '--month', nargs=1, type=int,
                          help='cron month parameter')
        cron.add_argument('-DW', '--dayofweek', nargs=1, type=int,
                          help='cron day of week parameter')
        # {type}
        group_enable = enable_parser.add_mutually_exclusive_group()
        group_enable.add_argument('--snapshot', action='store_true',
                                  help='Activate a snapshot configuration')
        group_enable.add_argument('--replica', action='store_true',
                                  help='Activate a replica configuration')
        # replica <args> options [-t <target volume>, -s serveur]
        enable_parser.add_argument('-t', '--target', nargs=1, type=str,
                                   help='zfs target pool name [default: None, '
                                   'same as source]')
        enable_parser.add_argument('-s', '--server', nargs=1, type=str,
                                   help='server hosting replication target '
                                   '[default: localhost]', required=False)
        # replica <args> type [--simple|--complete]
        replicag1 = enable_parser.add_mutually_exclusive_group()
        replicag1.add_argument('--simple', action='store_true',
                               help='replica type: simple, only target pool '
                               'and last dataset [default]', default=True)
        replicag1.add_argument('--complete', action='store_true',
                               help='replica type: complete, target pool plus '
                               'complete source dataset')
        # replica <args> archive [--latest|--oldest]
        replicag2 = enable_parser.add_mutually_exclusive_group()
        replicag2.add_argument('--latest', action='store_true',
                               help='replica archive: only latest snapshot '
                               '[default]', default=True)
        replicag2.add_argument('--oldest', action='store_true',
                               help='replica archive: complete set of snapshots')
        # last common <args>
        enable_parser.add_argument('volume', nargs=1, type=str,
                                   help='Mandatory zfs volume name')
        base = enable_parser.add_argument_group('common options')
        base.add_argument('-d', '--debug', action='store_true',
                          help='toggle debug [default: no]')
        base.add_argument('-l', '--log', action='store_true',
                          help='toggle logs [default: no]')
        args = enable_parser.parse_args(sys.argv[2:])
        self.arguments.update(vars(args))

    def disable(self):
        """ Parse the disable part of arguments """
        disable_parser = argparse.ArgumentParser(description='disable a configuration',
                                                 usage='%(prog)s disable {type} [options] <volume>',
                                                 add_help=True)
        group = disable_parser.add_mutually_exclusive_group()
        group.add_argument('--snapshot', action='store_true',
                           help='Deactivate a snapshot configuration')
        group.add_argument('--replica', action='store_true',
                           help='Deactivate a replica configuration')
        # last common <args>
        disable_parser.add_argument('volume', nargs=1, type=str,
                                    help='Mandatory zfs volume name')
        base = disable_parser.add_argument_group('common options')
        base.add_argument('-d', '--debug', action='store_true',
                          help='toggle debug [default: no]')
        base.add_argument('-l', '--log', action='store_true',
                          help='toggle logs [default: no]')
        args = disable_parser.parse_args(sys.argv[2:])
        self.arguments.update(vars(args))


def enable_snapshot():
    """ simple as snapshot """

    CONFIG2YAML['action'] = 'snapshot'

    # check and save configuration file
    check_and_save_yml()

    # then prepare and save crontab
    prepare_and_save_crontab()


def enable_replica():
    """ complicated as replica """

    CONFIG2YAML['action'] = 'replica'
    CONFIG2YAML.update({'replica': {'targetpool': '', 'server': ''}, })

    if script.arguments['server'] is None:
        CONFIG2YAML['replica']['server'] = 'localhost'
    else:
        CONFIG2YAML['replica']['server'] = str(script.arguments['server'][0])

    # target pool name, needed if server is localhost
    if script.arguments['target'] is None:
        if script.arguments['server'] is None:
            execo.log.logger.critical("Cannot replicate on same zfs pool.")
            sys.exit(1)
        else:
            CONFIG2YAML['replica']['targetpool'] = ''
    else:
        CONFIG2YAML['replica']['targetpool'] = str(script.arguments['target'][0])

    if script.arguments['complete']:
        # Include the dataset's properties in the stream. See 'man zfs'.
        CONFIG2YAML['replica']['type'] = 'complete'
    elif script.arguments['simple']:
        CONFIG2YAML['replica']['type'] = 'simple'

    if script.arguments['oldest']:
        # will replicate all available snapshots
        CONFIG2YAML['replica']['archive'] = 'oldest'
    elif script.arguments['latest']:
        # will replicate only the latest snapshot
        CONFIG2YAML['replica']['archive'] = 'latest'

    # check and save configuration file
    check_and_save_yml()

    # then prepare and save crontab
    prepare_and_save_crontab()


def disable_action(action=None):
    """ find and erase a configuration """

    nomfichier = ''.join([action, '-',
                          str(script.arguments['volume'][0]).replace('/', '-')])

    cronfichier = ''.join([GLOBAL_CONFIG['path']['cron'], nomfichier])

    if os.path.isfile(cronfichier):
        question = ' '.join(['Erase', str(cronfichier)])
        supprime = zfs_common.query_yesno(question)
        if supprime == 1:
            zfs_common.erase_file(cronfichier)

            conffichier = ''.join([GLOBAL_CONFIG['path']['conf'],
                                   nomfichier, '.yml'])

            if os.path.isfile(conffichier):
                question2 = ' '.join(['Erase', str(conffichier)])
                supprime2 = zfs_common.query_yesno(question2)

                if supprime2 == 1:
                    zfs_common.erase_file(conffichier)
            else:
                execo.log.logger.warning('file does not exist: ' + conffichier)

            restart_cron()
    else:
        execo.log.logger.warning('file does not exist: ' + cronfichier)


def check_and_save_yml():
    """ check and save configuration file """

    if script.arguments['snapshot']:
        nomfichier = '-'.join(['snapshot',
                               str(script.arguments['volume'][0]).replace('/', '-')])
        GLOBAL_CONFIG['path']['nomfichier'] = nomfichier

    elif script.arguments['replica']:
        nomfichier = '-'.join(['replica',
                               str(script.arguments['volume'][0]).replace('/', '-')])
        GLOBAL_CONFIG['path']['nomfichier'] = nomfichier

    tosave = ''.join([GLOBAL_CONFIG['path']['conf'],
                      GLOBAL_CONFIG['path']['nomfichier'], '.yml'])

    GLOBAL_CONFIG['path']['tosave'] = tosave  # needed later

    pprint(CONFIG2YAML)
    question = ' '.join(['Save this configuration:', tosave])
    verifconf = zfs_common.query_yesno(question)

    if verifconf == 1:
        zfs_common.verify_or_create_dir(tosave)
        zfs_common.save_yaml_file(tosave, CONFIG2YAML)
        execo.log.logger.info(str(tosave) + ' saved \n')
    else:
        execo.log.logger.info('nothing to do.')
        sys.exit(0)


def prepare_and_save_crontab():
    """ act as 'crontab -e' """

    # prepare crontab, all in str()
    minute = str(script.arguments['minute'][0])
    hour = str(script.arguments['hour'][0])
    if script.arguments['dayofmonth']:
        dom = str(script.arguments['dayofmonth'][0])
    else:
        dom = '*'
    if script.arguments['month']:
        month = str(script.arguments['month'][0])
    else:
        month = '*'
    if script.arguments['dayofweek']:
        dow = str(script.arguments['dayofweek'][0])
    else:
        dow = '*'

    # save crontab
    # deb12: introduce venv
    venv = 'cd GLOBAL_CONFIG['path']['tools'] && source bin/activate &&'
    binpy = ''.join(GLOBAL_CONFIG['path']['tools'],GLOBAL_CONFIG['path']['cmd'])
    ymlconf = GLOBAL_CONFIG['path']['tosave']
    global_cmd = ' '.join('(', venv, binpy, '-c', ymlconf, ')')

    if not script.arguments['mail']:
        cronlist = (minute, hour, dom, month, dow, GLOBAL_CONFIG['user'],
                    global_cmd, '>/dev/null')
        mailto = 'MAILTO=root'
    else:
        cronlist = (minute, hour, dom, month, dow, GLOBAL_CONFIG['user'],
                    global_cmd)
        mailto = 'MAILTO=' + str(script.arguments['mail'])

    crontab = (' '.join(cronlist))
    cronfichier = ''.join([GLOBAL_CONFIG['path']['cron'],
                           GLOBAL_CONFIG['path']['nomfichier']])
    print(mailto)
    print(crontab)
    question = ' '.join(['Save this crontab:', cronfichier])
    verifcron = zfs_common.query_yesno(question)
    if verifcron == 1:
        defaultpath = 'PATH=/sbin:/bin:/usr/sbin:/usr/bin'
        fichier = ['# ' + GLOBAL_CONFIG['date'], GLOBAL_CONFIG['defaultshell'],
                   mailto, defaultpath, crontab, '']
        zfs_common.verify_or_create_dir(cronfichier)
        zfs_common.save_crontab(cronfichier, fichier)
        execo.log.logger.info(cronfichier + ' saved \n')

        restart_cron()


def restart_cron():
    """ restart cron """
    question = zfs_common.query_yesno('Restart cron daemon')
    if question == 1:
        cmd = 'systemctl restart cron.service'
        zfs_common.execute_process(cmd)


if __name__ == '__main__':
    """
        verify admin rights
        load default config file
        set log level & log file
        parse config & args
        if enable:
            using args and config dict, prepare yml config dict
            ask user, then save config dict (as yaml file)
            prepare crontab file
            ask user, then save it
            ask user, restart cron daemon
        elif disable:
            using args and config dict, compute file's names and paths
            ask user, then erase crontab file
            ask user, then erase yaml config file
            ask user, restart cron daemon

    """
    # exit if not euid0
    zfs_common.verif_root()

    # read global config from a yml file
    GLOBAL_CONFIG = zfs_common.load_yaml_file(CONFIGFILE)
    GLOBAL_CONFIG['date'] = datetime.now().strftime('%Y/%m/%d-%H:%M:%S')

    # prepare configuration
    CONFIG2YAML = {'general': {}, }

    if GLOBAL_CONFIG['version'] != __version__:
        execo.log.logger.critical('Configuration version mismatch. Should be '
                                  'version ' + __version__)
        # TODO: explain 0.4 vs 0.5 differences in doc
        # Due to refactor, some options have changed and some have been merged.
        sys.exit(1)
    else:
        CONFIG2YAML['general']['version'] = __version__

    # parse CLI arguments
    script = ParseArgs()

    if script.arguments['debug']:
        CONFIG2YAML['general']['debug'] = 'yes'
        execo.log.logger.setLevel('DEBUG')
        execo.log.logger.debug(script.arguments)
        execo.log.logger.debug(GLOBAL_CONFIG)
    else:
        CONFIG2YAML['general']['debug'] = 'no'
        execo.log.logger.setLevel('INFO')

    if script.arguments['log']:
        CONFIG2YAML['general']['log'] = 'yes'
        logfile = ''.join([GLOBAL_CONFIG['path']['log'], 'zfs_actions.log'])
        CONFIG2YAML['general']['logfile'] = logfile
        file_handler = logging.FileHandler(logfile, 'a')
        # apply 'execo' formatter to file_handler
        file_handler.setFormatter(execo.logger.handlers[0].formatter)
        # create logpath before addHandler
        zfs_common.verify_or_create_dir(logfile)
        execo.log.logger.addHandler(file_handler)
    else:
        CONFIG2YAML['general']['log'] = 'no'
        # even if no log, specify the log file, one may change his mind
        logfile = ''.join([GLOBAL_CONFIG['path']['log'], 'zfs_actions.log'])
        CONFIG2YAML['general']['logfile'] = logfile

    CONFIG2YAML['volume'] = str(script.arguments['volume'][0])
    execo.log.logger.debug(CONFIG2YAML)

    if script.arguments['command'] == 'enable':
        execo.log.logger.debug('enable')

        if script.arguments['retention']:
            CONFIG2YAML['retention'] = str(script.arguments['retention'][0])
        else:
            CONFIG2YAML['retention'] = '7'

        if script.arguments['snapshot']:
            execo.log.logger.debug('snapshot')
            enable_snapshot()

        elif script.arguments['replica']:
            execo.log.logger.debug('replica')
            enable_replica()

    elif script.arguments['command'] == 'disable':
        execo.log.logger.debug('disable')

        if script.arguments['snapshot']:
            execo.log.logger.debug('snapshot')
            disable_action(action='snapshot')

        elif script.arguments['replica']:
            execo.log.logger.debug('replica')
            disable_action(action='replica')

    execo.log.logger.debug(GLOBAL_CONFIG)
    execo.log.logger.debug(CONFIG2YAML)
